### PR TITLE
Add ability to generate cpu and mem profiling

### PR DESCRIPTION
**Pull request by matheusd**
This adds the ability to generate the pprof cpu and memory profiles for the entire running time of the tool through config options `--cpu-profile` and `--mem-profile`.

These are useful to figure out hot paths for performance improvement when running the `check:` sub-comannds.

### CODE REVIEW

**Comment by heimdall-asguard** [comment]
If Heimdall is stuck, please click the following link: https://heimdall-public.cbhq.net/v1/pull_request/stuck?pr_number=145&repository=coinbase/rosetta-cli

**Comment by matheusd** [comment]
I can't access the CI environment so I'm not sure what's failing. Lint and tests are passing on my machine.

**Comment by patrick-ogrady** [comment]
@matheusd the `Salus` check is failing with this error: 
image

File: cmd/check_construction.go [hunk 1]
```diff
@@ -78,6 +78,8 @@ func runCheckConstructionCmd(cmd *cobra.Command, args []string) {
 			fmt.Errorf("%w: unable to initialize asserter", fetchErr.Err),
 			1,
 		)
+		cancel() // Avoid leaking context.
```
Comment (patrick-ogrady) [inline cmd/check_construction.go:81]: all these `Exit*` calls invoke `os.Exit` so we will never hit this

File: cmd/root.go [hunk 2]
```diff
@@ -45,8 +51,66 @@ var (
...
+// rootPreRun is executed before the root command runs and sets up cpu
+// profiling.
+func rootPreRun(*cobra.Command, []string) error {
+	if cpuProfile == "" {
+		return nil
+	}
+
+	f, err := os.Create(cpuProfile)
+	if err != nil {
+		return err
```
Comment (patrick-ogrady) [inline cmd/root.go:69]: Can we wrap these errors? `fmt.Errorf("%w: unable to create CPU Profile file"`)

File: cmd/root.go [hunk 3]
```diff
@@ -45,8 +51,66 @@ var (
...
 	SignalReceived = false
+
+	// profileCleanup is called after the root command is executed to
+	// cleanup a running cpu profile.
+	profileCleanup func()
 )
 
+// rootPreRun is executed before the root command runs and sets up cpu
+// profiling.
+func rootPreRun(*cobra.Command, []string) error {
```
Comment (patrick-ogrady) [inline cmd/root.go:62]: Would you mind linking in that this comes from: https://golang.org/pkg/runtime/pprof/#hdr-Profiling_a_Go_program

**Comment by patrick-ogrady** [comment]
@matheusd ... Love the ability to profile the `rosetta-cli`! I think you should revert the `exit.Code` related changes you made and we can discuss those in another PR. Happy to include the profiling changes in the next release.

**Comment by matheusd** [comment]
Reverting the `os.Exit` means the mem profile isn't generated because `rootPostRun` doesn't run after `os.Exit` is called.

But you do have a point about the exit status.

Here's an alternative: switch the `check:data` and `check:construction` commands to use `RunE` and then return an error on every code path that was resulting in an `os.Exit(1)` call. Then add the `os.Exit()` call at a single site, at the top level of the main package.

**Comment by patrick-ogrady** [comment]
> Reverting the os.Exit means the mem profile isn't generated because rootPostRun doesn't run after os.Exit is called.

Ahhhh my bad, yeah that makes sense.

> Here's an alternative: switch the check:data and check:construction commands to use RunE and then return an error on every code path that was resulting in an os.Exit(1) call. Then add the os.Exit() call at a single site, at the top level of the main package.

Let me put up a PR that cleans this up.

**Comment by matheusd** [comment]
Updated. I had to forgo using `PersistentPostRun` and hook the `rootPostRun()` directly on `Execute()` because when the subcommand fails, the corresponding `PersistentPostRun` isn't executed.

File: cmd/root.go [hunk 4]
```diff
@@ -45,18 +50,78 @@ var (
...
+	if cpuProfile == "" {
+		return nil
+	}
+
+	f, err := os.Create(cpuProfile)
+	if err != nil {
+		return err
+	}
+	if err := pprof.StartCPUProfile(f); err != nil {
+		f.Close()
```
Comment (patrick-ogrady) [inline cmd/root.go:71]: This is the unhandled error the `salus` test is failing on btw @matheusd

File: cmd/root.go [hunk 5]
```diff
@@ -45,18 +50,81 @@ var (
...
+//
+// Bassed on https://golang.org/pkg/runtime/pprof/#hdr-Profiling_a_Go_program
+func rootPreRun(*cobra.Command, []string) error {
+	if cpuProfile == "" {
+		return nil
+	}
+
+	f, err := os.Create(cpuProfile)
+	if err != nil {
+		return log.Printf("unable to create CPU profile file: %v", err)
```
Comment (patrick-ogrady) [inline cmd/root.go:70]: nit: this should be `fmt.Errorf("%w: unable to create CPU profile file", err)` instead of log.Printf?

Comment (patrick-ogrady) [reply cmd/root.go:70]: image

Comment (matheusd) [reply cmd/root.go]: No change is so small that one can't screw it up.... :roll_eyes:

Comment (patrick-ogrady) [reply cmd/root.go]: 🤣

**Review by patrick-ogrady — APPROVED** [review]

**Pull request merged**